Add a sidebar tab for documentation

AustinMroz opened the pull request:
Much belated, but I wanted to find solutions sufficiently clean for core.

Adds a new sidebar tab for displaying documentation information on nodes. Nodes utilizing the existing tooltip functionality will have their descriptions parsed. While this pane is open, hovering over items on the active node will instead scroll to and briefly highlight the corresponding item in the documentation pane. Alternatively, a node can supply a list for its `DESCRIPTION` property which consists of
- A short description which is used for node previews and title tooltips
- A full html description
- An optional dict to facilitate data transfer or provide additional configuration
  - `render`: A function called when the nodes documentation is displayed.
  - `select`: A function called when a node item has been hovered over.
  Note that both of these current options require a corresponding frontend extension to use. Few nodes will utilize this functionality, but it removes the need for ugly patching on those that do.

Further considerations
- Having NodeTooltips import the callback directly from documentationSidebar causes documentationSidebar to load before extensionManager is set in app. The current implementation allows this tooltipCallback to be hooked by other extensions, but seems the cleanest implementation.
- I quite dislike the question mark icon supplied by prime icons and implement an alternative with css. Mention was made of allowing icons other than those supplied by prime icons, but adding a generalized implementation here would be out of scope.

AustinMroz committed:
Initial sidebar documentation implementation

AustinMroz committed:
Remove unused styling code, unwrap

AustinMroz committed:
Formatting improvements,

The formatting of ndoes using the existing standardized tooltips has
been improved.

Experiemental work for assisting nodes with display of more detailed
descriptions

AustinMroz committed:
type implementation for detailed descriptions

Previously, description was a simple string, but supporting more complex
descriptions requires that new data be passed.

The type of a nodes description has been updated to be either a simple
string as before, or an array consisting of short description string, an
html string for the full description, and a placeholder dict for future
usage.

Definitions and usage points for description have been updated to
accommodate this change

AustinMroz committed:
Theming, pruning, and optional callbacks

Basic styling has been added to the display of documentation for nodes
using the existing tooltip system. This will need another pass to ensure
that style updates immediately when the light/dark toggle is hit instead
of requiring a change of node.

VHS specific namings have been replaced and the code for determining
what the mouse is hovering over has been removed. The existing tooltip
implementation is cleaner and will need to be integrated anyways so
tooltips are temporarily suppressed for the node actively being
displayed in the documentation sidebar.

Optional callbacks have been added for the initial sidebar display and a
user selecting a node element by hovering over it. While selection is
not yet implemented, this should cover any developer needs from more
involved collapsables to automated seeking to video timestamps.

AustinMroz committed:
Connect hover functionality, scroll fixes

Basic connecting for using the existing documentation hover code to
select an item from the active help pane.

Scrolling on selection will now properly perform the minium required
scroll to place the element on screen

AustinMroz committed:
Temporarily highlight item doc item on selection

AustinMroz committed:
Typing fixes, initial tests

AustinMroz committed:
Properly handle theme with css variables

AustinMroz committed:
Return styling to body, streamline tests

Styling was moved to the sidebar element for better organization, but
this caused errors when the new menu was not in use.

AustinMroz committed:
Move render callback to trigger on node change

AustinMroz committed:
Move css to style.css

Since the the css is now static the clutter of an added style element is
no longer needed

huchenlei commented:
> I quite dislike the question mark icon supplied by prime icons and implement an alternative with css. Mention was made of allowing icons other than those supplied by prime icons, but adding a generalized implementation here would be out of scope.

We do have iconify setup now. You can find from a large set of iconify icons here: https://icon-sets.iconify.design/. Here is an example usage in this repo:
https://github.com/Comfy-Org/ComfyUI_frontend/blob/39d68bcdc428751d0216e62fc6218b1555db1e46/src/components/graph/GraphCanvasMenu.vue#L36-L39

huchenlei commented on src/assets/css/style.css:
```diff
@@ -714,3 +714,24 @@ audio.comfy-audio.empty-audio-widget {
   padding: var(--comfy-tree-explorer-item-padding) !important;
 }
+
+.doc-node {
```
Please move this to corresponding vue component.

AustinMroz replied:
It took a good bit of fighting to make sure the template would actually update, but I agree do with migrating.
- Watching canvas is no longer viable after #1372, but for for the best. Discerning the best alternative was the primary blocker.

huchenlei commented on src/extensions/core/documentationSidebar.ts:
Please write this sidebar tab as a vue component similar to other sidebar tab components here: https://github.com/Comfy-Org/ComfyUI_frontend/tree/39d68bcdc428751d0216e62fc6218b1555db1e46/src/components/sidebar/tabs

huchenlei commented on src/types/apiTypes.ts:
```diff
@@ -344,6 +344,11 @@ const zComfyComboOutput = z.array(z.any())
 const zComfyOutputTypesSpec = z.array(
   z.union([zComfyNodeDataType, zComfyComboOutput])
 )
+const zDescriptionSpec = z.union([
```
Can you provide links to `DESCRIPTION` field used in custom nodes in this format?

AustinMroz replied:
A new format is being established here, so no node currently makes use of it. Node packs with a la carte documentation code currently need to play hot potato with swapping out the node def during load so that full descriptions aren't used for node previews or title hover tooltips. There should be a way for a node to describe documentation in place without requiring JS of any form, let alone one that requires abuse of the standards in place.

Test cases have been added for this format.

huchenlei commented on src/types/comfy.d.ts:
```diff
@@ -101,7 +101,7 @@ export interface ComfyExtension {
   name: string
   display_name?: string
-  description?: string
+  description?: [string | string, string | string, string, Record<string, any>]
```
`export type DescriptionSpec = z.infer<zDescriptionSpec>` in apiTypes.ts and use it here.

AustinMroz replied:
Much cleaner. Thanks.

huchenlei commented on src/components/graph/NodeTooltip.vue:
```diff
@@ -86,6 +86,11 @@ const onIdle = () => {
     ctor.title_mode !== LiteGraph.NO_TITLE &&
     canvas.graph_mouse[1] < node.pos[1] // If we are over a node, but not within the node then we are on its title
   ) {
+    if (comfyApp?.tooltipCallback?.(node, 'DESCRIPTION')) {
```
You can setup a pinia store to store/track the currently hovered node and node part instad of doing it via a callback. Setting up a pinia store will allow this state to be observable by multiple components if we need it later.

AustinMroz replied:
Done. The means of hiding tooltips for a node actively displayed in the documentation sidebar tab was concerning, but works better than expected.

huchenlei commented:
> We do have iconify setup now. You can find from a large set of iconify icons here: https://icon-sets.iconify.design/. Here is an example usage in this repo:

BTW I think I need to find a way support custom iconify icon. The iconify icon is lazy loaded during compile time so it won't be available for dynamic loaded extensions, but it should be necessary in this situation where in core we want to use iconify icon for sidebar tab icon.

AustinMroz committed:
Port sidebar documentation to vue component

huchenlei commented:
Related PR: https://github.com/Comfy-Org/ComfyUI_frontend/pull/1169

By making LGraphCanvas shallowReactive, we can observe root object's state directly, and hopefully have the active widget/slot logic handled by litegraph to reduce the overhead of additional event listening.

AustinMroz committed:
Update tooltip handling

AustinMroz committed:
Migrate to new sidebar registration

AustinMroz committed:
Functional node swaps under vue, icon update

AustinMroz committed:
Update tests for vue port

Mostly minor changes to selectors

Also fixes the glaringly obvious omission of the description field

AustinMroz committed:
Properly mirror the new description type

Remove errant logging

AustinMroz committed:
Migrate tooltip tracking to a pinia store

While I was concerned that doing this would remove the capability to
suppress tooltips on the active node, clearing the hoveredItem when it
used for documentation functions without even producing a temporary
tooltip.

A future commit will likely be made so that disabling tooltips for nodes
doesn't also prevent the hovered item from being tracked in the store.

AustinMroz committed:
Fix missing await

AustinMroz committed:
Add tests for advanced description formats

AustinMroz committed:
Merge sidebar-documentation onto main

The sidebar-documentation branch has diverged enough from main to
merging non-trivial

Remove deprecated type usage.
Move localization to language file.

AustinMroz committed:
Check for swapped node on interval.

Not ideal, but implementation is low cost and ensures the displayed
documentation properly updates.

AustinMroz committed:
Update tests

AustinMroz committed:
Prune dead code, tighten interval, add screenshot

Code related to possible options has been pruned. It was pointless for
updateNode to take a node as input. Collapsing documentation items does
not make sense given the greater space of a sidebar tab compared to a
floating window on the graph, so the corresponding code has been fully
pruned as well.

The topmost node is used instead of the current_node. While current_node
displayed the desired properties when canvas was still shallowReactive
of notifying a change in node, it's not intended to be used external to
litegraph and at best, tracks the topmost visible node instead of the
actual topmost node.

A test expectation screenshot has been added for verifying theming works
for the documentation tab.

huchenlei reviewed (changes requested):
Can you take a look at the failing tests?

huchenlei commented:
Closing for inactivity.
